dbcon/mysql/rulebased_optimizer.cpp

```diff
@@ -0,0 +1,189 @@
+/* Copyright (C) 2025 MariaDB Corporation
+
+   This program is free software; you can redistribute it and/or
+   modify it under the terms of the GNU General Public License
+   as published by the Free Software Foundation; version 2 of
+   the License.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program; if not, write to the Free Software
+   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+   MA 02110-1301, USA. */
+
+#include "execplan/calpontselectexecutionplan.h"
+#include "execplan/simplecolumn.h"
+#include "rulebased_optimizer.h"
+
+namespace optimizer {
+
+static const std::string RewrittenSubTableAliasPrefix = "$added_sub_";
+
+// Apply a list of rules to a CSEP
+bool optimizeCSEPWithRules(execplan::CalpontSelectExecutionPlan& root, const std::vector<Rule>& rules) {
+
+  bool changed = false;
+  for (const auto& rule : rules)
+  {
+    changed |= rule.apply(root);
+  }
+  return changed;
+}
+
+// high level API call for optimizer
+bool optimizeCSEP(execplan::CalpontSelectExecutionPlan& root)
+{
+  optimizer::Rule parallelCES{"parallelCES", optimizer::matchParallelCES, optimizer::applyParallelCES};
+  
+  std::vector<Rule> rules = {parallelCES};
+
+  return optimizeCSEPWithRules(root, rules);
+}
+
+// Apply iteratively until CSEP is converged by rule
+bool Rule::apply(execplan::CalpontSelectExecutionPlan& root) const
+{
+  bool changedThisRound = false;
+  bool hasBeenApplied = false;
+  do
+  {
+    changedThisRound = walk(root) && !applyOnlyOnce;
+    hasBeenApplied = changedThisRound;
+  } while (changedThisRound);
+
+  return hasBeenApplied;
+}
+
+// DFS walk to match CSEP and apply rules if match
+bool Rule::walk(execplan::CalpontSelectExecutionPlan& csep) const
+{
+  bool rewrite = false;
+
+  for (auto& table : csep.derivedTableList())
+  {
+    auto* csepPtr = dynamic_cast<execplan::CalpontSelectExecutionPlan*>(table.get());
+    if (!csepPtr)
+    {
+      continue;
+    }
+
+    auto& csepLocal = *csepPtr;
+    rewrite |= walk(csepLocal);
```

Comment:
nit: Personally, I'm not comfortable using recursion in c++. Maybe rewrite it as an iterative traversal?